### PR TITLE
Rename and stabilize useEditorFeature as useSetting

### PR DESCRIPTION
**Pull request by youknowriad**
Related #31416 

This is an important API for Global Settings (theme.json) APIs

### CODE REVIEW

**Comment by oandregal** [comment]
I'm not stoked with the name `useThemeSetting`, I think is misleading as the setting can actually come from core, theme, or user. :thinking: The current name is fine by me. Alternatives names I can brainstorm are: `useEditorSetting`, `useSetting`, etc.

**Comment by youknowriad** [comment]
Fair,

 - `useEditorSetting` I don't like it because we're already on "editor", so it's kind of redundant.
 - `useSetting` I like that one but the issue I have with it is that the `settings` of the block-editor package include more than just the "theme settings", so folks using the package might be wondering why it returns the theme settings but not the other settings (say mediaUpload handler...)

I think this particular API should be thought more in relation to the block-editor package semantics and not WordPress semantics,  in other words, whatever name we choose here is the same one that should be used to pass these settings to the block editor provider `<BlockEditorProvider settings={ { __experimentalFeatures: something, ...otherKindOfSettings } } />`

Right now, we use `__experimentalFeatures` there, but at some point we need to change that and stabilize it (maybe not in 5.8 though), so the question is: what do these settings actually represent for a block-editor instance (could be the post editor, could be a standalone block editor...)?

I think it's a bit hard to reply to this question but right now I can't think of anything better than `themeSettings`, also the final format could be `theme: { settings , styles }` (It doesn't matter for the block editor instance where the values come from, whether they are customized, persisted or just come from a file)

**Comment by oandregal** [comment]
> I think this particular API should be thought more in relation to the block-editor package semantics and not WordPress semantics.
> (...)
> what do these settings actually represent for a block-editor instance (could be the post editor, could be a standalone block editor...)?

I think those are great frames of mind for this. I see two paths forward:

- Make this API retrieve _any_ setting present in the store => `getSetting`?
- Make this API retrieve _only_ settings that can be set/unset by block (that's the only difference `__experimentalFeatures` has with the others) => `getBlockSetting`?

> I think it's a bit hard to reply to this question but right now I can't think of anything better than themeSettings, also the final format could be theme: { settings , styles }

Isn't this at odds with separating the block-editor package from WordPress semantics suggested above?

Independently from what direction/name we chose for this API, I was thinking that this data could be structured in the store in this way:

- the top-level settings of theme.json are retrofit to the existing settings (or create new ones) ― [see](https://github.com/WordPress/gutenberg/issues/30082#issuecomment-816679803), no longer [unset them](https://github.com/WordPress/gutenberg/blob/trunk/lib/global-styles.php#L132).
- we create a new key in that store for per-block settings (`__experimentalFeatures` becomes `blocks` and has only block data)

**Comment by youknowriad** [comment]
I'm not sure we should make this API retrieve any block-editor setting because setting a "media uploader" or a "link fetcher" or "reusable blocks" is not something that can be mixed with "allowing colors" or "allowed image sizes".

> Make this API retrieve only settings that can be set/unset by block (that's the only difference __experimentalFeatures has with the others) => getBlockSetting?

Not sure I understand this.. The block can't set/unset a setting?

> Isn't this at odds with separating the block-editor package from WordPress semantics suggested above?

No in the sense that I think that a "theme" is something that can have meaning in any block-editor. "Theme" might not be the right word though but it's the best one that I can think of right now.

> Moving these to the top level

Yes maybe, it does make `useSetting` compelling but I'm also very unsure about  mixing these settings with the examples I have above.

---
So it seems we have two options:

 - Have a subkey, call it `theme` or not.

To give more context to the discussion, here are some of the top level settings that we already have. Do we think that these settings can make sense as well in a `useSetting` hook?

```
 * @property {boolean} alignWide Enable/Disable Wide/Full Alignments
 * @property {boolean} supportsLayout Enable/disable layouts support in container blocks.
 * @property {boolean} imageEditing Image Editing settings set to false to disable.
 * @property {Array} imageSizes Available image sizes
 * @property {number} maxWidth Max width to constraint resizing
 * @property {boolean|Array} allowedBlockTypes Allowed block types
 * @property {boolean} hasFixedToolbar Whether or not the editor toolbar is fixed
 * @property {boolean} focusMode Whether the focus mode is enabled or not
 * @property {Array} styles Editor Styles
 * @property {boolean} keepCaretInsideBlock Whether caret should move between blocks in edit mode
 * @property {string} bodyPlaceholder Empty post placeholder
 * @property {string} titlePlaceholder Empty title placeholder
 * @property {boolean} codeEditingEnabled Whether or not the user can switch to the code editor
 * @property {boolean} __experimentalCanUserUseUnfilteredHTML Whether the user should be able to use unfiltered HTML or the HTML should be filtered e.g., to remove elements considered insecure like iframes.
 * @property {boolean} __experimentalBlockDirectory Whether the user has enabled the Block Directory
 * @property {Array} __experimentalBlockPatterns Array of objects representing the block patterns
 * @property {Array} __experimentalBlockPatternCategories Array of objects representing the block pattern categories
 ```

**Comment by oandregal** [comment]
Sorry, I messed up the words. I meant:

> Make this API retrieve only settings whose value can be set per block (that's the only difference `__experimentalFeatures` has with the other settings) => getBlockSetting?

**Comment by youknowriad** [comment]
Renamed to just useSetting per the discussion here :)

**Comment by oandregal** [comment]
To make sure I follow: if we go with `getSetting`, does it mean that we expect that function (in [packages/edit-site/src/components/editor/utils.js](packages/edit-site/src/components/editor/utils.js) and [packages/block-editor/src/components/use-setting/index.js](https://github.com/WordPress/gutenberg/pull/31587/files?file-filters%5B%5D=.js#diff-f3dffb5f90413eb8cfb4d350f403e8a844e6293b2bd7e4b8cfc6209518e6a306R76)) to fallback to `settings` in the store if the path is not found in `settings.__experimentalFeatures` (first block, then top-level)? This is: `useSetting('alignWide')` or `useSetting('imageEditing')`should be possible.

I'm fine if we do that in a follow-up PR, just want to confirm I understand the direction.

**Comment by youknowriad** [comment]
> This is: useSetting('alignWide') or useSetting('imageEditing')should be possible.

Maybe later, it doesn't matter right now, I expect over time most of these settings will disappear in favor of theme.json related ones. Also the internal implementation of the hook is not important for me.

**Comment by youknowriad** [comment]
I do think thought that for 5.8, `__experimentalFeatures` should stay experimental, moving these to top level make sense but we should do it carefully.

**Comment by oandregal** [comment]
> Maybe later, it doesn't matter right now, I expect over time most of these settings will disappear in favor of theme.json related ones.

So the idea is, at the moment, only expose things configured via theme.json. I see how that can be confusing for consumers in the interim, but I also understand that there's still a mechanism to query the whole settings object (the store's selector used by the hook), so we should be fine. The important part to keep stable for consumers is the path (`color.palette`, `color.custom`, etc).

**Review by oandregal — APPROVED** [review]

File: packages/block-editor/src/components/use-setting/index.js [hunk 1]
```diff
@@ -50,19 +50,19 @@ const deprecatedFlags = {
 };
 
 /**
- * Hook that retrieves the setting for the given editor feature.
+ * Hook that retrieves the editor setting.
```
Comment (oandregal) [inline packages/block-editor/src/components/use-setting/index.js:53]: Should we clarify this to state it's not for all settings, but https://github.com/WordPress/gutenberg/pull/31587#issuecomment-838382549 ?

**Pull request merged**